### PR TITLE
Feature/597 collapse messages and parameters

### PR DESCRIPTION
**Pull request by neoxelox**
Related: https://github.com/latitude-dev/latitude-llm/issues/597

- In the playground preview, chat and document log info you now can collapse parameters and individual messages.
- In the document log info parameters used for that run are now shown.
- Bring the fixes for the PromptL compiler to the legacy one, so everyone can enjoy collapsing their parameters :) . https://github.com/latitude-dev/promptl/pull/6
- Fixes bug that crashed the document editor page when putting a non text content in system messages or an image content in an assistant message.

### CODE REVIEW

File: packages/core/src/services/ai/providers/rules/index.ts [hunk 1]
```diff
@@ -44,5 +45,19 @@ export function applyCustomRules({
...
 
+  return rules
+}
+
+export function applyVercelRules({
+  providerType,
+  messages,
+  config,
+}: Props): AppliedRules {
+  let rules: AppliedRules = {
```
Comment (andresgutgon) [inline packages/core/src/services/ai/providers/rules/index.ts:56]: `const` please

File: packages/core/src/services/ai/providers/rules/vercel.ts [hunk 2]
```diff
@@ -23,7 +23,7 @@ function flattenSystemMessage({
 }): Message[] {
   const content = message.content as TextContent[] | string
 
-  // NOTO: `applyCustomRules` can be invoked multiple times
+  // NOTO: `applyVercelRules` can be invoked multiple times
```
Comment (andresgutgon) [inline packages/core/src/services/ai/providers/rules/vercel.ts:26]: `s/NOTO/NOTE/g`

**Review by csansoon — APPROVED** [review]

**Review by andresgutgon — DISMISSED** [review]

File: packages/compiler/src/compiler/compile.ts [hunk 3]
```diff
@@ -158,35 +159,80 @@ export class Compile {
...
+
+    text = text.trimEnd()
+    sourceMap = sourceMap.map((ref) => ({
+      ...ref,
+      start: Math.min(ref.start, text.length),
+      end: Math.min(ref.end, text.length),
+    }))
+
+    return sourceMap
+  }
```
Comment (geclos) [inline packages/compiler/src/compiler/compile.ts:208]: is this to support collapsing variables in old compiler? Honestly... i wouldn't have done it 😂 I guess I'm too late for that

File: packages/web-ui/src/ds/molecules/Chat/Message/index.tsx [hunk 4]
```diff
@@ -58,30 +76,117 @@ export function Message({
...
+      )
+    }
+    const nextSegment = value.slice(
+      sourceMap[i]!.end,
+      sourceMap[i + 1]?.start ?? value.length,
+    )
+    if (nextSegment.length > 0) segments.push(nextSegment)
+  }
+
+  // Group segments by new line
```
Comment (geclos) [inline packages/web-ui/src/ds/molecules/Chat/Message/index.tsx:156]: instead of a comment put it in a function with an easy-to-understand name

File: packages/web-ui/src/ds/molecules/Chat/Message/index.tsx [hunk 5]
```diff
@@ -58,30 +76,117 @@ export function Message({
...
+          currentGroup = []
+        }
+      }
+    } else {
+      currentGroup.push(segment)
+    }
+  }
+  if (currentGroup.length > 0) groups.push(currentGroup)
+
+  // Render groups
```
Comment (geclos) [inline packages/web-ui/src/ds/molecules/Chat/Message/index.tsx:175]: ai comments are so useless sometimes

File: packages/web-ui/src/ds/molecules/Chat/Message/index.tsx [hunk 6]
```diff
@@ -58,30 +76,117 @@ export function Message({
...
+          <div className='line-clamp-6'>{sourceMap[i]!.identifier!}</div>
+        </Tooltip>,
+      )
+    }
+    const nextSegment = value.slice(
+      sourceMap[i]!.end,
+      sourceMap[i + 1]?.start ?? value.length,
+    )
+    if (nextSegment.length > 0) segments.push(nextSegment)
+  }
```
Comment (geclos) [inline packages/web-ui/src/ds/molecules/Chat/Message/index.tsx:154]: tell the ai to refactor for readability there's lots of code that can be put in chunks here

Comment (geclos) [inline packages/web-ui/src/ds/molecules/Chat/Message/index.tsx:154]: this whole for function smells like another method

**Review by geclos — DISMISSED** [review]

File: packages/web-ui/src/ds/molecules/Chat/Message/index.tsx (hunk 5 above)
Comment (neoxelox) [reply packages/web-ui/src/ds/molecules/Chat/Message/index.tsx:175]: not AI lol

Comment (andresgutgon) [reply packages/web-ui/src/ds/molecules/Chat/Message/index.tsx:175]: But it is kind of useless, tho

File: packages/core/src/services/ai/index.ts [hunk 7]
```diff
@@ -94,12 +94,23 @@ export async function ai({
...
+    if (rule.rules.length > 0) {
+      return Result.error(
+        new ChainError({
+          code: RunErrorCodes.AIRunError,
+          message:
+            'There are rule violations:\n' +
+            rule.rules.map((rule) => `- ${rule.ruleMessage}`).join('\n'),
+        }),
+      )
+    }
```
Comment (andresgutgon) [inline packages/core/src/services/ai/index.ts:112]: Why is this necessary now? How were the rules returned before when some rule failed here?

File: packages/core/src/services/ai/providers/rules/custom.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
+import { ContentType, MessageRole } from '@latitude-data/compiler'
+
+import { AppliedRules, ProviderRules } from '.'
+
+export function applyCustomRules(rules: AppliedRules): AppliedRules {
```
Comment (andresgutgon) [inline packages/core/src/services/ai/providers/rules/custom.ts:5]: What means custom?

**Review by andresgutgon — APPROVED** [review]

File: packages/core/src/services/ai/index.ts (hunk 7 above)
Comment (neoxelox) [reply packages/core/src/services/ai/index.ts:112]: We never checked for rules before, so it crashed
